### PR TITLE
[BEAM-10234] Create ApproximateDistinct using HLL Impl

### PR DESCRIPTION
**Pull request by rezarokni**
[BEAM-10234] Create ApproximateDistinct using HLL Impl
Create ApproximateDistinct which uses HLL Impl
Marked ApproximateUnique as deprecated in preparation for removal in favour of ApproximateDistinct. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by rezarokni** [comment]
@robinyqiu

**Comment by rezarokni** [comment]
@zfraa

File: sdks/java/extensions/zetasketch/src/main/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinct.java [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+ * <pre>{@Code
+ *     p.apply("Int", Create.of(ints)).apply("IntHLL", ApproximateCountDistinct.globally()
+ *     .withPercision(PRECISION));
+ *     }</pre>
+ *
+ * <h4>Example 2: Approximate Count of Key Value {@code PCollection<KV<Integer,Foo>>}</h4>
+ *
+ * <pre>{@code
+ * PCollection<KV<Integer, Long>> result =
+ * p.apply("Int", Create.of(ints)).apply("LongHLL", ApproximateCountDistinct.perKey());
```
Comment (robinyqiu) [inline sdks/java/extensions/zetasketch/src/main/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinct.java:64]: This loc does not match description above

File: sdks/java/extensions/zetasketch/src/main/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinct.java [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+ * }</pre>
+ *
+ * <h4>Example 3: Approximate Count of Key Value {@code PCollection<KV<Integer,Foo>>}</h4>
+ *
+ * <pre>{@code
+ * PCollection<KV<Integer, Foo>> approxResultInteger =
+ *          p.apply("Int", Create.of(Foo))
+ *          .apply("IntHLL", ApproximateCountDistinct.<Integer, KV<Integer, Integer>>perKey()
+ *          .via(x -> KV.of(x.getKey(), (long) x.hashCode())));
```
Comment (robinyqiu) [inline sdks/java/extensions/zetasketch/src/main/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinct.java:73]: I think the function should be something like `kv -> KV.of(kv.getKey(), kv.getValue().hashcode())`?

File: sdks/java/extensions/zetasketch/src/test/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinctTest.java [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+
+    PCollection<Long> approxResultInteger =
+        p.apply(
+                "Int",
+                Create.of(
+                    INTS1.stream().map(x -> KV.of(x, KV.of(x, x))).collect(Collectors.toList())))
+            .apply(
+                "IntHLL",
+                ApproximateCountDistinct.<KV<Integer, KV<Integer, Integer>>>globally()
+                    .via((KV<Integer, KV<Integer, Integer>> x) -> (long) x.getValue().hashCode()));
```
Comment (robinyqiu) [inline sdks/java/extensions/zetasketch/src/test/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinctTest.java:268]: I think the function here should be 'x -> x.hashCode()' without 'getValue'?

File: sdks/java/extensions/zetasketch/src/test/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinctTest.java [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+        ints.add(KV.of(i, KV.of(i, k)));
+      }
+    }
+
+    PCollection<KV<Integer, Long>> approxResultInteger =
+        p.apply("Int", Create.of(ints))
+            .apply(
+                "IntHLL",
+                ApproximateCountDistinct.<Integer, KV<Integer, Integer>>perKey()
+                    .via(x -> KV.of(x.getKey(), (long) x.hashCode()))
```
Comment (robinyqiu) [inline sdks/java/extensions/zetasketch/src/test/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinctTest.java:293]: Same here. Shouldn't we use `x -> KV.of(x.getKey(), x.getValue().hashCode())` here?

**Review by robinyqiu — COMMENTED** [review]
Thank you so much for working on this Reze! Happy to see this finally will be available to customers. Generally LGTM. Left some comments.

Also maybe we will want to update Changes.md to announce this new feature.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ApproximateUnique.java [hunk 5]
```diff
@@ -56,7 +56,10 @@
  *
  * For more details about using {@code HllCount} and the {@code zetasketch} extension module, see
  * https://s.apache.org/hll-in-beam#bookmark=id.v6chsij1ixo7.
+ *
+ * @deprecated use {@code ApproximateCountDistinct} which has moved to HLL implementation.
```
[the remark this answers is not in the document]
Comment (rezarokni) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ApproximateUnique.java:60]: Changed the comments. But not added snippets yet, would it be ok to just assume they will look at the new location?

File: sdks/java/extensions/zetasketch/src/main/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinct.java [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+ *
+ * <p>We make use of the {@link HllCount} implementation for this transform. Please use {@link
+ * HllCount} directly if you need access to the sketches.
+ *
+ * <p>If the object is not one of {@link Byte[]} {@link Integer} {@link Double} {@link String} make
+ * use of {@link Globally#via} or {@link PerKey#via}
+ *
+ * <h3>Examples</h3>
+ *
+ * <h4>Example 1: Approximate Count of Longs {@code PCollection<Long>} and specify precision</h4>
```
[the remark this answers is not in the document]
Comment (rezarokni) [reply sdks/java/extensions/zetasketch/src/main/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinct.java:53]: Done.

File: sdks/java/extensions/zetasketch/src/main/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinct.java [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+ *
+ * <pre>{@Code
+ *     p.apply("Int", Create.of(ints)).apply("IntHLL", ApproximateCountDistinct.globally()
```
[the remark this answers is not in the document]
Comment (rezarokni) [reply sdks/java/extensions/zetasketch/src/main/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinct.java:56]: Yes, corrected.

File: sdks/java/extensions/zetasketch/src/main/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinct.java [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+      public abstract Globally<T> build();
+    }
+
+    public Globally<T> via(ProcessFunction<T, Long> fn) {
+
+      return toBuilder().setMapping(Contextful.<T, Long>fn(fn)).build();
+    }
+
+    public <V> Globally<V> withPercision(Integer withPercision) {
+      this.toBuilder().setPrecision(withPercision).build();
```
[the remark this answers is not in the document]
Comment (rezarokni) [reply sdks/java/extensions/zetasketch/src/main/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinct.java:133]: Thanx.

File: sdks/java/extensions/zetasketch/src/main/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinct.java (hunk 1 above)
Comment (rezarokni) [reply sdks/java/extensions/zetasketch/src/main/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinct.java:64]: Done.

File: sdks/java/extensions/zetasketch/src/test/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinctTest.java (hunk 3 above)
Comment (rezarokni) [reply sdks/java/extensions/zetasketch/src/test/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinctTest.java:268]: The value in this case is the KV<int,int> we are using a KV here as a convenient object , so we need the unique from the value which comes from KV<int,KV<int,int>>

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ApproximateUnique.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (robinyqiu) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ApproximateUnique.java]: I am ok with that.

File: sdks/java/extensions/zetasketch/src/test/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinctTest.java (hunk 3 above)
Comment (robinyqiu) [reply sdks/java/extensions/zetasketch/src/test/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinctTest.java:268]: I see. This tests will pass, but I find it a bit hard to understand. How about we use `x -> x.getValue().getValue()`? I find that simpler to understand. (e.g. KV<x, KV<x, x>> will be identified by x)

**Review by robinyqiu — APPROVED** [review]
LGTM

**Comment by rezarokni** [comment]
Sorry @robinyqiu I accidentally clicked on re-review and cant see a way to undo the click!

File: sdks/java/extensions/zetasketch/src/test/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinctTest.java (hunk 3 above)
Comment (rezarokni) [reply sdks/java/extensions/zetasketch/src/test/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinctTest.java:268]: If we getValue().getValue() then we will get the value of the 2nd KV which is not going to be unique as KV<1,KV<2,3>> is not == to KV<1,KV<3,3>> . If KV is an issue I can create a new POJO just for this test. Using a KV within a KV is just an easy convenience.

**Comment by rezarokni** [comment]
Run Java PreCommit

File: sdks/java/extensions/zetasketch/src/test/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinctTest.java (hunk 3 above)
Comment (rezarokni) [reply sdks/java/extensions/zetasketch/src/test/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinctTest.java:268]: PS as you have LGTM is it ok to merge as is , or do you need the test changed before ?

Comment (rezarokni) [reply sdks/java/extensions/zetasketch/src/test/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinctTest.java:268]: @robinyqiu are you ok with this being merged without the change to the Test ? Thanx.

Comment (robinyqiu) [reply sdks/java/extensions/zetasketch/src/test/java/org/apache/beam/sdk/extensions/zetasketch/ApproximateCountDistinctTest.java:268]: I am ok with merging it as it is.

**Comment by rezarokni** [comment]
Run Java PreCommit

**Pull request merged**